src/bdb.c

```diff
@@ -23,162 +23,120 @@ void usage(void)
+                                // when you start using the state.
+
     fprintf(f, "%s ", inst_name(i->type));
     if (inst_has_operand(i->type))
     {
-        if (i->type == INST_NATIVE)
-        {
-            fprintf(f, SV_Fmt, SV_Arg(state->native_labels[i->operand.as_u64]));
-        }
-        else if (i->type == INST_CALL || i->type == INST_JMP || i->type == INST_JMP_IF)
-        {
-            fprintf(f, SV_Fmt, SV_Arg(state->labels[i->operand.as_u64]));
-        }
-        else
-        {
-            fprintf(f, "%" PRIu64, i->operand.as_i64);
-        }
+        fprintf(f, "%" PRIu64, i->operand.as_i64);
     }
 }
 
-void bdb_add_breakpoint(Bdb_State *state, Inst_Addr addr)
+Bdb_Breakpoint *bdb_find_breakpoint_by_addr(Bdb_State *bdb, Inst_Addr addr)
 {
-    assert(state);
-
-    if (addr > state->bm.program_size)
-    {
-        fprintf(stderr, "ERR : Symbol out of program\n");
-        return;
+    for (size_t i = 0; i < bdb->breakpoints_size; ++i) {
+        if (bdb->breakpoints[i].addr == addr) {
+            return &bdb->breakpoints[i];
+        }
     }
 
-    if (addr > BM_PROGRAM_CAPACITY)
-    {
-        fprintf(stderr, "ERR : Symbol out of memory\n");
-        return;
-    }
+    return NULL;
+}
 
-    if (state->breakpoints[addr].is_enabled)
-    {
+void bdb_add_breakpoint(Bdb_State *state, Inst_Addr addr, String_View label)
+{
+    assert(state);
+    assert(state->breakpoints_size < BDB_BREAKPOINTS_CAPACITY);
+
+    if (bdb_find_breakpoint_by_addr(state, addr)) {
         fprintf(stderr, "ERR : Breakpoint already set\n");
         return;
     }
 
-    state->breakpoints[addr].is_enabled = 1;
+    state->breakpoints[state->breakpoints_size].is_broken = 0;
+    state->breakpoints[state->breakpoints_size].label =
+        arena_sv_dup(&state->break_arena, label); // NOTE: moving the label to an arena
+                                                  // with longer lifetime
+    state->breakpoints[state->breakpoints_size].addr = addr;
+    state->breakpoints_size += 1;
+
+    fprintf(stdout, "INFO : Breakpoint set at %"PRIu64"\n", addr);
 }
 
 void bdb_delete_breakpoint(Bdb_State *state, Inst_Addr addr)
 {
     assert(state);
 
-    if (addr > state->bm.program_size)
-    {
-        fprintf(stderr, "ERR : Symbol out of program\n");
+    Bdb_Breakpoint *breakpoint = bdb_find_breakpoint_by_addr(state, addr);
+    if (breakpoint == NULL) {
+        fprintf(stderr, "ERR : No such breakpoint\n");
         return;
     }
 
-    if (addr > BM_PROGRAM_CAPACITY)
-    {
-        fprintf(stderr, "ERR : Symbol out of memory\n");
-        return;
-    }
+    assert(breakpoint <= state->breakpoints);
 
-    if (!state->breakpoints[addr].is_enabled)
-    {
-        fprintf(stderr, "ERR : No such breakpoint\n");
-        return;
+    state->breakpoints_size -= 1;
+    memmove(breakpoint,
+            breakpoint + 1,
+            state->breakpoints_size - (size_t) (state->breakpoints - breakpoint));
```

Comment:
This size calculation is probably incorrect. Size should be in bytes, but in here it's most probably in amount of breakpoints.